File: cpp/Solved/BOJ/11780 플로이드 2.cpp
#include <bits/stdc++.h>
using namespace std;
// 11780 플로이드 2

int floyd[101][101];
int nxt[101][101];
int main(void) {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int n, m;
    cin >> n >> m;
    n++;
    for (int j = 0; j < m; j++) {
        int a, b, c;
        cin >> a >> b >> c;
        floyd[a][b] = (floyd[a][b]) ? min(floyd[a][b], c) : c;
        nxt[a][b] = b;
    }
    for (int k = 1; k < n; k++) {
        for (int i = 1; i < n; i++) {
            if (i == k) continue;
            for (int j = 1; j < n; j++) {
                if (j == k) continue;
                if (i == j) continue;
                int bypass = floyd[i][k] + floyd[k][j];
                if (floyd[i][j] && floyd[i][k] && floyd[k][j]) {
                    if (bypass < floyd[i][j]) {
                        floyd[i][j] = bypass;
                        nxt[i][j] = nxt[i][k];
                    }
                } else if (floyd[i][k] && floyd[k][j]) {
                    floyd[i][j] = bypass;
                    nxt[i][j] = nxt[i][k];
                }
            }
        }
    }
    for (int i = 1; i < n; i++) {
        for (int j = 1; j < n; j++) {
            cout << floyd[i][j] << ' ';
        }
        cout << '\n';
    }
    for (int i = 1; i < n; i++) {
        for (int j = 1; j < n; j++) {
            if (i == j) {
                cout << 0 << '\n';
                continue;
            }
            vector<int> route = {i};
            while (1) {
                if (route.back() == j) {
                    cout << route.size() << ' ';
                    for (auto r : route) cout << r << ' ';
                    cout << '\n';
                    break;
                } else if (route.back() == 0) {
                    cout << 0 << '\n';
                    break;
                }
                route.push_back(nxt[route.back()][j]);
            }
        }
    }
}

File: Python3/Solved/BOJ/11780 플로이드 2.py
# 11780 플로이드 2
input = __import__('sys').stdin.readline

n, m = int(input()), int(input())
n += 1
mxN = 10**8+1
floyd = [[mxN]*(n) for _ in range(n)]
route = [[0]*(n) for _ in range(n)]

for _ in range(m):
    a, b, c = map(int, input().split())
    floyd[a][b] = min(floyd[a][b], c)
    route[a][b] = b

for k in range(1, n):
    for i in range(1, n):
        if i == k: continue
        for j in range(1, n):
            if j == k: continue
            if i == j: floyd[i][j] = 0
            bypass = floyd[i][k] + floyd[k][j]
            if bypass < floyd[i][j]:
                floyd[i][j] = bypass
                route[i][j] = route[i][k]

for i in range(n):
    for j in range(n):
        if floyd[i][j] == mxN: floyd[i][j] = 0

[print(' '.join(map(str, f[1:]))) for f in floyd[1:]]
for i in range(1, n):
    for j in range(1, n):
        if i==j:
            print(0)
            continue
        nxt = [i]
        while 1:
            if nxt[-1] == j:
                print(len(nxt), ' '.join(map(str, nxt)))
                break
            elif nxt[-1] == 0:
                print(0)
                break
            nxt.append(route[nxt[-1]][j])